### PR TITLE
Bugfix for paths for Synced Realms

### PR DESCRIPTION
**Pull request by edualonso**

### CODE REVIEW

File: packages/cinterop/src/commonMain/kotlin/io/realm/internal/interop/RealmInterop.kt [hunk 1]
```diff
@@ -234,6 +234,7 @@ expect object RealmInterop {
     fun realm_app_log_out(app: RealmAppPointer, user: RealmUserPointer, callback: AppCallback<Unit>)
     fun realm_app_remove_user(app: RealmAppPointer, user: RealmUserPointer, callback: AppCallback<Unit>)
     fun realm_clear_cached_apps()
+    fun realm_app_sync_client_get_default_file_path_for_realm(app: RealmAppPointer, syncConfig: RealmSyncConfigurationPointer, name: String): String
```
Comment (cmelchior) [inline packages/cinterop/src/commonMain/kotlin/io/realm/internal/interop/RealmInterop.kt:237]: Isn't `name` optional here?

File: packages/library-base/src/commonMain/kotlin/io/realm/Configuration.kt [hunk 2]
```diff
@@ -142,54 +140,20 @@ public interface Configuration {
...
-         *      NSUserDomainMask,
-         *      null,
-         *      true,
-         *      null
-         * )
-         * ```
+         * If setting the full path of the realm, this name is not taken into account.
          *
-         * @param directoryPath either the canonical absolute path or a relative path from the current directory ('./').
+         * @throws IllegalAttributeException if the name includes a path separator.
```
Comment (cmelchior) [inline packages/library-base/src/commonMain/kotlin/io/realm/Configuration.kt:147]: What is `IllegalAttributeException`? Shouldn't this be `IllegalArgumentException`?

File: packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/SyncConfiguration.kt [hunk 3]
```diff
@@ -31,6 +33,8 @@ import io.realm.mongodb.internal.SyncConfigurationImpl
 import io.realm.mongodb.internal.UserImpl
 import kotlin.reflect.KClass
 
+private const val REALM_EXTENSION = ".realm"
```
Comment (cmelchior) [inline packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/SyncConfiguration.kt:36]: Perhaps rename to `REALM_FILE_EXTENSION`?

File: packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/SyncConfiguration.kt [hunk 4]
```diff
@@ -99,13 +111,22 @@ public interface SyncConfiguration : Configuration {
...
-         *
-         * @param errorHandler lambda to handle the error.
-         */
-        public fun errorHandler(errorHandler: SyncSession.ErrorHandler): Builder =
-            apply { this.errorHandler = errorHandler }
+        override fun name(name: String): Builder = apply {
+            if (name.contains(PATH_SEPARATOR)) {
+                throw IllegalArgumentException("Name cannot contain path separator '$PATH_SEPARATOR': '$name'")
+            }
+            if (name.isEmpty()) {
```
Comment (cmelchior) [inline packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/SyncConfiguration.kt:118]: This check is missing in `RealmConfiguration`

**Review by cmelchior — COMMENTED** [review]
I know this is still draft, so just leaving a few drive-by comments.

File: packages/cinterop/src/commonMain/kotlin/io/realm/internal/interop/RealmInterop.kt (hunk 1 above)
Comment (edualonso) [reply packages/cinterop/src/commonMain/kotlin/io/realm/internal/interop/RealmInterop.kt:237]: Yes, I missed it. I will change it.

File: test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncConfigTests.kt [hunk 5]
```diff
@@ -242,6 +243,26 @@ class SyncConfigTests {
         assertEquals(config.partitionValue.asString(), partitionValue)
     }
 
+    @Test
+    fun name() {
```
Comment (cmelchior) [inline test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncConfigTests.kt:247]: I think we lack some corner case tests here:

- E.g. partition-based default names. They should be something like `s_mypartition.realm` or `i_42.realm`.
- What happens if you bring a filename with no `.realm` suffix. 

Since our default name is `default.realm` and if ObjectStore doe append `.realm` to all names, then I could be tempted to also enforce it for local Realms, but that would be a breaking change. Not sure what your thoughts are?

It does have the advantage that you can always double-click on a file and get it to open in Realm Studio though.

File: test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncSessionTests.kt [hunk 6]
```diff
@@ -65,20 +63,18 @@ class SyncSessionTests {
         user = runBlocking {
             app.createUserAndLogIn(email, password)
         }
-        tmpDir = PlatformUtils.createTempDir()
```
Comment (cmelchior) [inline test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncSessionTests.kt:68]: Are you sure it is safe to just not use the temporary dir? How is cleanup done then? I would have assumed that the `TestApp` then needs to define this? Or we have some other logic that deletes the default folder after each test?

**Review by cmelchior — CHANGES_REQUESTED** [review]
Lots of nice improvements in here 💯 .... I left some minor comments, but also a few bigger ones:

- Naming between SyncConfiguration and RealmConfiguration seems to be inconsistent with regard to the `.realm` suffix.
- We lack tests for corner cases around naming.
- Missing CHANGELOG. This PR is breaking paths for synched builds, so we should make sure to mention it. Also directory is no longer available for SyncConfiguration and `AppConfiguration.Builder.syncRootDirectory` has been added.

File: packages/cinterop/src/darwin/kotlin/io/realm/internal/interop/RealmInterop.kt [hunk 7]
```diff
@@ -1440,6 +1436,20 @@ actual object RealmInterop {
...
+        syncConfig: RealmSyncConfigurationPointer,
+        name: String?
+    ): String {
+        val cPath = realm_wrapper.realm_app_sync_client_get_default_file_path_for_realm(
+            app.cptr(),
+            syncConfig.cptr(),
+            name
+        )
+        return cPath.safeKString()
+            .also { realm_wrapper.realm_free(cPath) }
```
[the remark this answers is not in the document]
Comment (edualonso) [reply packages/cinterop/src/darwin/kotlin/io/realm/internal/interop/RealmInterop.kt:1450]: Not that I'm aware of. I did it because the C-API docs explicitly state the following:
```
/**
 * Get the default realm file path based on the user and partition value in the config.
 *
 * @param custom_filename custom name for the realm file itself. Can be null,
 *                        in which case a default name based on the config will be used.
 *
 * Return value must be manually released with realm_free().
 */
```
There are other C-API functions that require it but we haven't implemented them yet.

[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/cinterop/src/darwin/kotlin/io/realm/internal/interop/RealmInterop.kt:1450]: :+1: Sounds like we need to go through all of them then 🙈

File: packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/AppConfiguration.kt [hunk 8]
```diff
@@ -106,6 +113,36 @@ public interface AppConfiguration {
                 this.userLoggers = customLoggers
             }
 
+        /**
+         * Configures the root folder containing all files and realms used when synchronizing data
+         * between the device and MongoDB Realm.
+         *
+         * The default root directory is platform-dependent:
+         * - Android: `Context.getFilesDir()/mongodb-realm`
```
[the remark this answers is not in the document]
Comment (edualonso) [reply packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/AppConfiguration.kt:121]: Yeah, I forgot about adding the extra sync part.

File: packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/AppConfiguration.kt [hunk 9]
```diff
@@ -106,6 +113,36 @@ public interface AppConfiguration {
...
+        public fun syncRootDirectory(rootDir: String): Builder = apply {
+            val directoryExists = directoryExists(rootDir)
+            if (!directoryExists && fileExists(rootDir)) {
+                throw IllegalArgumentException("'rootDir' is a file, not a directory: $rootDir.")
+            }
+            if (!directoryExists) {
+                try {
+                    prepareRealmDirectoryPath(rootDir)
+                } catch (e: IllegalArgumentException) {
+                    throw IllegalArgumentException("Could not create the specified directory: $rootDir.")
```
[the remark this answers is not in the document]
Comment (edualonso) [reply packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/AppConfiguration.kt:137]: `prepareRealmDirectoryPath` already handles the cases of `!dir.mkdirs()` and `dir.isFile` internally. This catch will only be triggered if `File(directoryPath, filename).absolutePath` fails for some reason (not related to Realm) so I guess it's OK to not throw this exception manually.

File: packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/SyncConfiguration.kt [hunk 10]
```diff
@@ -99,13 +114,17 @@ public interface SyncConfiguration : Configuration {
...
-         *
-         * @param errorHandler lambda to handle the error.
-         */
-        public fun errorHandler(errorHandler: SyncSession.ErrorHandler): Builder =
-            apply { this.errorHandler = errorHandler }
+        override fun name(name: String): Builder = apply {
+            nameCheck(name)
+
+            // Strip `.realm` suffix as it will be appended by Object Store later
+            this.name = if (name.endsWith(REALM_FILE_EXTENSION)) {
```
[the remark this answers is not in the document]
Comment (edualonso) [reply packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/SyncConfiguration.kt:121]: Yeps, `realm_app_sync_client_get_default_file_path_for_realm` does append `.realm` to whatever `name` we use.

File: test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncConfigTests.kt (hunk 5 above)
Comment (edualonso) [reply test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncConfigTests.kt:247]: I think it would be fair to add the constraint for local Realms for the sake of uniformity.

File: test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncSessionTests.kt (hunk 6 above)
Comment (edualonso) [reply test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncSessionTests.kt:68]: Yeah, I noticed that too. I sort of left it in limbo while I was working on implementing the rest. I will sort it out.

Comment (edualonso) [reply test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncSessionTests.kt:68]: Turns out cleanup is done correctly when we call `TestApp.close()`. Also and just for fun I changed the deletion logic so that we use `PlatformUtils.deleteTempDir` instead of OKIO's `FileSystem.deleteRecursively` and it also works fine. If I'm honest, I don't recall why we reached the conclusion that we needed to use OKIO 😑  although after this PR it might make sense to use it if we keep having the need for handling more IO scenarios.

File: CHANGELOG.md [hunk 11]
```diff
@@ -4,6 +4,7 @@
 * [Sync] `SyncConfiguration` and `SyncSession` has been moved to `io.realm.mongodb.sync`.
 * [Sync] `EmailPasswordAuth` has been movedto `io.realm.mongodb.auth`.
 * [Sync] Improved exception hiearchy for App and Sync exceptions. All sync/app exceptions now use `io.realm.mongodb.exceptions.AppException` as their top-level exception type. Many methods have more specialized exceptions for common errors that can be caught and reacted to. See `AppException` documentation for more details.
+* [Sync] `SyncConfiguration.directory` is no longer available and `AppConfiguration.syncRootDirectory` has been added instead to allow users to set the root folder containing all files used for data synchronization between the device and MongoDB Realm. (Issue [#795](https://github.com/realm/realm-kotlin/issues/795))
```
Comment (cmelchior) [inline CHANGELOG.md:7]: ``SyncConfiguration.directory` is no longer available` this should move to `Breaking Changes`

File: packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/AppConfiguration.kt [hunk 12]
```diff
@@ -106,6 +113,47 @@ public interface AppConfiguration {
...
             }
 
+        /**
+         * Configures the root folder containing all files and realms used when synchronizing data
+         * between the device and MongoDB Realm.
+         *
+         * The default root directory is platform-dependent:
+         * ```
+         * // For Android the default directory is obtained using
+         * val dir = ${Context.getFilesDir()/mongodb-realm}
```
Comment (cmelchior) [inline packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/AppConfiguration.kt:123]: Nitpick...but this isn't actually valid Kotlin code...these should be encapsulated in `""` for that.

File: test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncConfigTests.kt [hunk 13]
```diff
@@ -242,25 +241,74 @@ class SyncConfigTests {
...
-//    @Test
-//    fun name_illegalValuesThrows() {
-//        val user: User = createTestUser(app)
-//        val builder = SyncConfiguration.Builder(user, DEFAULT_PARTITION)
-//
-//        assertFailsWith<IllegalArgumentException> { builder.name(TestHelper.getNull()) }
-//        assertFailsWith<IllegalArgumentException> { builder.name(".realm") }
-//    }
+    @Test
+    fun name_withoutFileExtension() {
```
Comment (cmelchior) [inline test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncConfigTests.kt:245]: We are missing a test for what the default value would look like.

**Review by cmelchior — COMMENTED** [review]
Nice...just left a few comments. Will do a final review once the name changes around the file extension we discussed has landed

File: test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncConfigTests.kt (hunk 13 above)
Comment (edualonso) [reply test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncConfigTests.kt:245]: I added `name_similarToDefaultObjectStoreName` below, if that's what you mean?

File: CHANGELOG.md (hunk 11 above)
Comment (edualonso) [reply CHANGELOG.md:7]: I'm not sure what you mean by moving this to breaking changes. This is the breaking changes section.

I will move the part about `AppConfiguration.syncRootDirectory` to the `Fixed` section.

File: packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/AppConfiguration.kt (hunk 12 above)
Comment (edualonso) [reply packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/AppConfiguration.kt:123]: 🙈

File: test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncConfigTests.kt (hunk 13 above)
Comment (cmelchior) [reply test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncConfigTests.kt:245]: Yes, but that also passes in a user-defined name here: https://github.com/realm/realm-kotlin/pull/800/files/2c8370b257770e55e75b92ee8e731b6afb380939#diff-f3172b56e038676ca6061b7181e7eaae31885fe86018ccc332e00d2274b3524dR297

We are missing a test for what happens if no name is given.

File: CHANGELOG.md (hunk 11 above)
Comment (cmelchior) [reply CHANGELOG.md:7]: Ups 🙈 ... for some reason I thought this was under Enhancements.

File: test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncConfigTests.kt (hunk 13 above)
Comment (edualonso) [reply test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncConfigTests.kt:245]: All the other tests in the suite that are not name-related don't specify a name so that should already be covered, no? On top of that, if Object Store assigns somewhat "random" names internally, to which don't have access until we instantiate the config object itself, what is it that we should test? I don't see how to compare the resulting name to an expected value.

Comment (cmelchior) [reply test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SyncConfigTests.kt:245]: But we do know....partition values would be encoded with something like `s_<value>.realm` for Strings `i_<value>.realm` for Ints

The reason we need would to be ensure it was mostly for testing that the suffix `.realm` appear for these as well.

**Review by cmelchior — APPROVED** [review]
Awesome :shipit: ... I had a few minor comments to tests and documentation but addressed them in a commit instead.

This can be merged if CI is happy.
